### PR TITLE
Memory efficient fragmentation

### PR DESCRIPTION
**Pull request by rmillikin**

### CODE REVIEW

File: EngineLayer/Indexing/IndexingEngine.cs [hunk 1]
```diff
@@ -109,21 +109,20 @@ protected override MetaMorpheusEngineResults RunSpecific()
...
                 {
-                    globalPeptides.AddRange(localPeptides);
+                    peptides.AddRange(localPeptides);
                 }
             });
 
             // sort peptides by mass
-            List<PeptideWithSetModifications> peptidesSortedByMass = globalPeptides.OrderBy(p => p.MonoisotopicMass).ToList();
-            globalPeptides = null; //RAM
+            peptides.Sort((x, y) => x.MonoisotopicMass.CompareTo(y.MonoisotopicMass));
```
Comment (zrolfs) [inline EngineLayer/Indexing/IndexingEngine.cs]: neat!

File: EngineLayer/Indexing/IndexingEngine.cs [hunk 2]
```diff
@@ -233,7 +234,7 @@ private List<int>[] CreateNewPrecursorIndex(List<PeptideWithSetModifications> pe
                     }
                 }
                 progress++;
-                var percentProgress = (int)((progress / maxProgress) * 100);
```
Comment (zrolfs) [inline EngineLayer/Indexing/IndexingEngine.cs]: This line should be "int percentProgress = (int)((progress*1d/maxProgress)*100);"
In its current form, the percentProgress can only ever be 0 or 1.

File: EngineLayer/MetaMorpheusEngine.cs [hunk 3]
```diff
@@ -100,7 +101,7 @@ public static List<MatchedFragmentIon> MatchFragmentIons(Ms2ScanWithSpecificMass
 
                     if (commonParameters.ProductMassTolerance.Within(scan.TheScan.MassSpectrum.XArray[closestMzIndex], theoreticalFragmentMz))
                     {
-                        matchedFragmentIons.Add(new MatchedFragmentIon(product, theoreticalFragmentMz, scan.TheScan.MassSpectrum.YArray[closestMzIndex], 1));
+                        matchedFragmentIons.Add(new MatchedFragmentIon(ref product, theoreticalFragmentMz, scan.TheScan.MassSpectrum.YArray[closestMzIndex], 1));
```
Comment (zrolfs) [inline EngineLayer/MetaMorpheusEngine.cs]: Isn't the product already a reference type and this is redundant?

File: EngineLayer/MetaMorpheusEngine.cs [hunk 4]
```diff
@@ -128,16 +131,17 @@ public static List<MatchedFragmentIon> MatchFragmentIons(Ms2ScanWithSpecificMass
                 // is the mass error acceptable?
                 if (commonParameters.ProductMassTolerance.Within(closestExperimentalMass.monoisotopicMass, product.NeutralMass) && closestExperimentalMass.charge <= scan.PrecursorCharge)
                 {
-                    matchedFragmentIons.Add(new MatchedFragmentIon(product, closestExperimentalMass.monoisotopicMass.ToMz(closestExperimentalMass.charge),
+                    matchedFragmentIons.Add(new MatchedFragmentIon(ref product, closestExperimentalMass.monoisotopicMass.ToMz(closestExperimentalMass.charge),
```
Comment (zrolfs) [inline EngineLayer/MetaMorpheusEngine.cs]: same here

File: EngineLayer/MetaMorpheusEngine.cs [hunk 5]
```diff
@@ -152,7 +156,7 @@ public static List<MatchedFragmentIon> MatchFragmentIons(Ms2ScanWithSpecificMass
                     // is the mass error acceptable?
                     if (commonParameters.ProductMassTolerance.Within(closestExperimentalMass.monoisotopicMass, compIonMass) && closestExperimentalMass.charge <= scan.PrecursorCharge)
                     {
-                        matchedFragmentIons.Add(new MatchedFragmentIon(product, closestExperimentalMass.monoisotopicMass.ToMz(closestExperimentalMass.charge),
+                        matchedFragmentIons.Add(new MatchedFragmentIon(ref product, closestExperimentalMass.monoisotopicMass.ToMz(closestExperimentalMass.charge),
```
Comment (zrolfs) [inline EngineLayer/MetaMorpheusEngine.cs]: lotta these, huh?

File: EngineLayer/MetaMorpheusEngine.cs (hunk 3 above)
Comment (rmillikin) [reply EngineLayer/MetaMorpheusEngine.cs]: `Product` is a struct now

**Comment by zrolfs** [comment]
This doesn't quite hit the code coverage. You could add a test that has a DeadEndNH2 and a test where either the alpha or beta peptides (but not both) has null for its protein.

**Comment by zrolfs** [comment]
That merge screwed up a lot of stuff. I'm working on it right now.

File: EngineLayer/CrosslinkSearch/CrosslinkedPeptides.cs [hunk 6]
```diff
@@ -67,7 +73,9 @@ public class CrosslinkedPeptide
             List<int> modPos, PeptideWithSetModifications peptide)
         {
             Dictionary<Tuple<int, int>, List<Product>> AllTheoreticalFragmentsLists = new Dictionary<Tuple<int, int>, List<Product>>();
-            var originalFragments = peptide.Fragment(dissociationType, FragmentationTerminus.Both).ToList();
+            var originalFragments = new List<Product>();
+            peptide.Fragment(dissociationType, FragmentationTerminus.Both, originalFragments);
+            var l = new List<Product>();
```
Comment (lonelu) [inline EngineLayer/CrosslinkSearch/CrosslinkedPeptides.cs]: 'l' is not a good name.

File: Test/XLTest.cs [hunk 7]
```diff
@@ -803,16 +811,16 @@ public static void TestMs2Ms3()
                 csm.BetaPeptide.ResolveAllAmbiguities();
             }
             // test parent scan (CID)
-            Assert.That(csm.MatchedFragmentIons.Count == 37);
+            Assert.AreEqual(csm.MatchedFragmentIons.Count, 32);
             Assert.That(csm.ScanNumber == 2);
 
             // test child scan (low-resolution CID, alpha peptide signature ion)
-            Assert.That(csm.ChildMatchedFragmentIons.First().Key == 6);
```
Comment (zrolfs) [inline Test/XLTest.cs]: @lonelu can you look at this? I don't feel good about this after the merge conflicts.

Comment (zrolfs) [reply Test/XLTest.cs]: It's almost like the alpha and beta peptides switched?

Comment (zrolfs) [reply Test/XLTest.cs]: After looking at this, the alpha and beta peptides switched BEFORE the merge, so this might actually be fine. Concerning that the new fragmentation caused changes, but I think the merge "Merge branch 'master' into memoryEfficientFragmentation 0644682" is OK.

Comment (rmillikin) [reply Test/XLTest.cs]: #1780 resolved some bugs in the XL search, I'm guessing that fixing those caused some scores to change

**Comment by zrolfs** [comment]
I checked out the current version of this and compared it to the current master on a HeLa/trypsin file.
The two classic searches produced identical numbers of psms/peptides/proteins at 1% FDR. This PR brought the total search time down from 184 sec to 145. That's awesome.

**Comment by zrolfs** [comment]
That was with a fasta. I'll try again with an xml.

**Comment by zrolfs** [comment]
XML Results on HeLa/Trypsin: Still identical psms/peptides/proteins. Time: 217s -> 157s
